### PR TITLE
Grpc gateway, createUser, GetUsers and Auth Service (only generating token)

### PR DESCRIPTION
**Pull request by estebanalarcon1globant**

### CODE REVIEW

File: proto/user.proto [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  rpc Authenticate (AuthReq) returns (AuthResp) {
+    option (google.api.http) = {
+      post: "/v1/login"
+      body: "*"
+    };
+  }
+}
+
+message CreateUserReq {
```
Comment (jcromanu) [inline proto/user.proto:30]: Acá se recomienda intercalar los indices del proto como 1 , 3 , 5 etc en lugar de secuenciarlos 1, 2, 3 ,4

File: users/domain/user_repository.go [hunk 2]
```diff
@@ -5,4 +5,5 @@ import "context"
 type UserRepository interface {
 	CreateUser(ctx context.Context, user Users) (int, error)
 	GetUsers(ctx context.Context, limit int, offset int) ([]Users, error)
+	GetUserByEmail(_ context.Context, email string) (Users, error)
```
Comment (jcromanu) [inline users/domain/user_repository.go]: aunque la interfaz sea casi explicita , nombrar los campos como ctx

File: main.go [hunk 3]
```diff
@@ -3,31 +3,51 @@ package main
...
+}
+
+func (c *Car) Run() {
+	fmt.Println("Car is running")
+}
+
+func (t *Transport) Run() {
+	fmt.Println("Transport is running")
+}
+
```
Comment (javibauza) [inline main.go]: Remove this block. I think we used it for testing purposes.

File: users/repository/user_repository_test.go [hunk 4]
```diff
@@ -116,3 +118,31 @@ func (s *SuiteUser) TestUserRepository_GetUsers() {
...
+
+	email := "test@test.com"
+	errExpected := errors.New("record not found")
+
+	s.mock.ExpectQuery("SELECT (.+) FROM `users` WHERE (.+)").
+		//WithArgs(sqlmock.AnyArg(), sqlmock.AnyArg()).
+		WillReturnError(errExpected)
+	_, errGot := s.repository.GetUserByEmail(context.TODO(), email)
+	require.EqualError(s.T(), errExpected, errGot.Error())
+}
```
Comment (javibauza) [inline users/repository/user_repository_test.go:148]: Investigate table-driven tests.

File: users/service/user_service.go [hunk 5]
```diff
@@ -1,31 +1,98 @@
...
+
+	// retrieve user with pwd from database
+	user, err := s.GetUserByEmail(ctx, auth.Email)
+	if err != nil {
+		return "", err
+	}
+
+	// verify password
+	if user.PwdHash != utils.HashSHA256(auth.Password) {
+		return "", errors.New("invalid credentials")
```
Comment (javibauza) [inline users/service/user_service.go]: Maybe you should use a custom error package.

File: users/repository/user_repository_test.go (hunk 4 above)
Comment (javibauza) [inline users/repository/user_repository_test.go:148]: Investigate table-driven tests.

File: users/transport/grpc.go [hunk 6]
```diff
@@ -1,79 +1,96 @@
...
-	createUser gt.Handler
-	getUsers   gt.Handler
+	pb.UnimplementedUserServiceServer
+	createUser   gt.Handler
+	getUsers     gt.Handler
+	authenticate gt.Handler
 }
 
 var (
 	ErrBadRequest = errors.New("bad request. Doesn't match interface")
```
Comment (javibauza) [inline users/transport/grpc.go]: Consider using a custom error package.

File: utils/encrypt.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+import (
+	"crypto/sha1"
+	"encoding/base64"
+)
+
+func HashSHA256(stringToHash string) string {
+	hasher := sha1.New()
+	hasher.Write([]byte(stringToHash))
+	return base64.URLEncoding.EncodeToString(hasher.Sum(nil))
+}
```
Comment (javibauza) [inline utils/encrypt.go:12]: Consider using bcrypt for encryption.

File: users/domain/user_repository.go (hunk 2 above)
Comment (estebanalarcon1globant) [reply users/domain/user_repository.go]: ok corregido

File: main.go (hunk 3 above)
Comment (estebanalarcon1globant) [reply main.go]: removed

File: users/service/user_service.go (hunk 5 above)
Comment (estebanalarcon1globant) [reply users/service/user_service.go]: yes. I created another file with all custom errors

File: users/transport/grpc.go (hunk 6 above)
Comment (estebanalarcon1globant) [reply users/transport/grpc.go]: ready

File: utils/encrypt.go (hunk 7 above)
Comment (estebanalarcon1globant) [reply utils/encrypt.go:12]: Im gonna read about bcrypt for the next implementations

File: users/repository/user_repository_test.go (hunk 4 above)
Comment (estebanalarcon1globant) [reply users/repository/user_repository_test.go:148]: okay

**Review by javibauza — APPROVED** [review]

File: proto/user.proto (hunk 1 above)
Comment (estebanalarcon1globant) [reply proto/user.proto:30]: entiendo, ya cambie las secuencias

**Review by javibauza — APPROVED** [review]

**Pull request merged**